server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java

```diff
@@ -0,0 +1,162 @@
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(mimeMessage)
+            .attribute(new Attribute(OWNER_ATTRIBUTE_NAME, AttributeValue.of(new SerializableUser(deletedMessage.getOwner()))))
+            .attribute(new Attribute(HAS_ATTACHMENT_ATTRIBUTE_NAME, AttributeValue.of(deletedMessage.hasAttachment())))
+            .attribute(new Attribute(DELIVERY_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeliveryDate()))))
+            .attribute(new Attribute(DELETION_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeletionDate()))))
+            .attribute(new Attribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME, AttributeValue.of(serializedMailboxIds)))
+            .attribute(new Attribute(SUBJECT_ATTRIBUTE_VALUE, subjectAttributeValue))
+            .build();
+    }
+
+    DeletedMessage fromMail(Mail mail) {
+        MemoizedSupplier<byte[]> byteContent = new MemoizedSupplier<>(Throwing.supplier(
+            () -> MimeMessageUtil.asBytes(mail.getMessage())));
+
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(ImmutableList.copyOf(mail.getRecipients()))
+            .content(() -> new ByteArrayInputStream(byteContent.get()))
+            .hasAttachment(retrieveHasAttachment(mail))
+            .subject(retrieveSubject(mail))
+            .build();
+
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
+                .map(Attribute::getValue)
+                .map(AttributeValue::getValue)
+                .filter(value -> value instanceof Optional)
+                .map(value -> (Optional) value)
+                .map(optional -> optional.map(obj -> {
+                    Preconditions.checkArgument(obj instanceof String, "mail should have a 'subject' attribute being of type 'Optional<String>");
+                    return (String) obj;
+                }))
+                .orElseThrow(() -> new IllegalArgumentException("mail should have a 'subject' attribute being of type 'Optional<String>"));
+    }
+
+    private boolean retrieveHasAttachment(Mail mail) {
+        Attribute attribute = mail.getAttribute(HAS_ATTACHMENT_ATTRIBUTE_NAME).orElseThrow(() -> new IllegalArgumentException("mail should have a 'hasAttachment' attribute"));
+        Preconditions.checkArgument(attribute.getValue().value() instanceof Boolean, "mail should have a 'hasAttachment' attribute of type boolean");
+        return (boolean) (Boolean) attribute.getValue().value();
```

Comment:
that's definitely too ugly for me

Reply:
Sorry, I did not write it consciously....